crypto/crmf/crmf_lib.c

```diff
@@ -0,0 +1,769 @@
+    }
+
+    if ((tmpl->extensions) == NULL) {
+        if ((tmpl->extensions = sk_X509_EXTENSION_new_null()) == NULL)
+            goto oom;
+        new = 1;
+    }
+
+    if (!sk_X509_EXTENSION_push(tmpl->extensions, (X509_EXTENSION *)ext))
+        goto oom;
+    return 1;
+ oom:
+    CRMFerr(CRMF_F_OSSL_CRMF_MSG_PUSH0_EXTENSION, ERR_R_MALLOC_FAILURE);
+
+    if (new != 0) {
+        sk_X509_EXTENSION_free(tmpl->extensions);
+        tmpl->extensions = NULL;
+    }
+    return 0;
+}
+
+/*
+ * TODO: also support cases 1+2 defined in RFC4211, section 4.1.
+ * returns pointer to created OSSL_CRMF_POPOSIGNINGKEY on success, NULL on error
+ */
+static OSSL_CRMF_POPOSIGNINGKEY *CRMF_poposigkey_new(OSSL_CRMF_CERTREQUEST *cr,
+                                                     const EVP_PKEY *pkey,
+                                                     int dgst)
+{
+    OSSL_CRMF_POPOSIGNINGKEY *ps = NULL;
+    int l;
+    size_t crlen, max_sig_size;
+    unsigned int siglen;
+    unsigned char *crder = NULL, *sig = NULL;
+    int alg_nid=0, md_nid=0;
+    const EVP_MD *alg = NULL;
+
+    EVP_MD_CTX *ctx = NULL;
+
+    if (cr == NULL || pkey == NULL) {
+        CRMFerr(CRMF_F_CRMF_POPOSIGKEY_NEW, CRMF_R_NULL_ARGUMENT);
+        return 0;
+    }
+
+    if ((ps = OSSL_CRMF_POPOSIGNINGKEY_new()) == NULL)
+        goto err;
+
+    /* OpenSSL defaults all bit strings to be encoded as ASN.1 NamedBitList */
+    ps->signature->flags &= ~(ASN1_STRING_FLAG_BITS_LEFT | 0x07);
+    ps->signature->flags |= ASN1_STRING_FLAG_BITS_LEFT;
+
+    l = i2d_OSSL_CRMF_CERTREQUEST(cr, &crder);
+    if (l < 0 || crder == NULL)
+        goto err;
+    crlen = (size_t) l;
+
+    max_sig_size = EVP_PKEY_size((EVP_PKEY *)pkey);
+    sig = OPENSSL_malloc(max_sig_size);
+    if (sig == NULL)
+        goto err;
+
+    if (!OBJ_find_sigid_by_algs(&alg_nid, dgst, EVP_PKEY_id(pkey))) {
+        CRMFerr(CRMF_F_CRMF_POPOSIGKEY_NEW,
+                CRMF_R_UNSUPPORTED_ALG_FOR_POPSIGNINGKEY);
+        goto err;
+    }
+    if (!(OBJ_find_sigid_algs(alg_nid, &md_nid, NULL) &&
+          (alg = EVP_get_digestbynid(md_nid)) != NULL)) {
+        CRMFerr(CRMF_F_CRMF_POPOSIGKEY_NEW,
+                CRMF_R_UNSUPPORTED_ALG_FOR_POPSIGNINGKEY);
+        goto err;
+    }
+    X509_ALGOR_set0(ps->algorithmIdentifier, OBJ_nid2obj(alg_nid),
+                    V_ASN1_NULL, NULL);
+
+    ctx = EVP_MD_CTX_create();
+    if (!(EVP_SignInit_ex(ctx, alg, NULL)))
+        goto err;
+    if (!(EVP_SignUpdate(ctx, crder, crlen)))
+        goto err;
+    if (!(EVP_SignFinal(ctx, sig, &siglen, (EVP_PKEY *)pkey)))
+        goto err;
+
+    if (!(ASN1_BIT_STRING_set(ps->signature, sig, siglen)))
+        goto err;
+
+    /* cleanup */
+    OPENSSL_free(crder);
+    EVP_MD_CTX_destroy(ctx);
+    OPENSSL_free(sig);
+    return ps;
+ err:
+    CRMFerr(CRMF_F_CRMF_POPOSIGKEY_NEW, CRMF_R_ERROR);
+    OSSL_CRMF_POPOSIGNINGKEY_free(ps);
+    OPENSSL_free(crder);
+    if (ctx)
+        EVP_MD_CTX_destroy(ctx);
```

Comment:
EVP_MD_CTX_create() and EVP_MD_CTX_destroy() were renamed to
EVP_MD_CTX_new() and EVP_MD_CTX_free() in OpenSSL 1.1.0.

Reply:
Updated